src/NBitcoin/NetworkOptions.cs

```diff
@@ -0,0 +1,149 @@
+﻿namespace NBitcoin
+{
+    /// <summary>
+    /// The current use-case for this class is to provide option-dependent serialization.
+    /// It is a drop-in replacement for the legacy TransactionOptions enumeration.
+    /// 
+    /// The NetworkOptions instance is intended to be relayed from the Network object to 
+    /// the child objects: Network => Block => Transaction => TxIn/TxOut => Coin => ...
+    /// It can also be used on repository or stream objects that need to deserialize the 
+    /// above objects or passed as a parameter futher down the call-hierarchy as needed.
+    /// </summary>
+    public class NetworkOptions
+    {
+        public const uint None = 0x00000000;
+        public const uint Witness = 0x40000000;
+        public const uint All = Witness;
+
+        //TODO?: Could be used by Block:
+        //public Action<NetworkOptions, Block> SetBlockSpecificFlags { get; set; } = null;
+        //TODO?: Could be used by Transaction:
+        //public Action<NetworkOptions, Transaction> SetTransactionSpecificFlags { get; set; } = null;
+
+        private uint flags = All;
+
+        /// <summary>
+        /// Creates a NetworkOptions object.
+        /// </summary>
+        /// <param name="flags">The flags to maintain in the object.</param>
+        public NetworkOptions(uint flags = All)
+        {
+            this.flags = flags;
+        }
+
+        /// <summary>
+        /// Clones the NetworkOptions object.
+        /// </summary>
+        /// <returns>The cloned NetworkOptions object.</returns>
+        public NetworkOptions Clone()
+        {
+            var clone = new NetworkOptions();
+            clone.flags = this.flags;
+            return clone;
+        }
+
+        /// <summary>
+        /// Allows this object to be explicitly cast to a uint that contains the flags.
+        /// </summary>
+        /// <param name="o">The NetworkOptions object to cast.</param>
+        public static explicit operator uint(NetworkOptions o)
+        {
+            return o?.flags ?? All;
```

Comment:
Are you sure the default should be `all` and not `none`?

Reply:
All = Witness. So either it should be All or Witness. I have found from PR 816 that when passing TransactionOptions as a optional argument it always made sense to put the default as TransactionOptions.All. For the class-based solution the compiler is not accepting NetworkOptions.All as a default argument so we will have to pass null instead. Then we definitely need null to default to All/Witness.